components/expenses/DeleteExpenseButton.js

```diff
@@ -0,0 +1,110 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { Mutation } from '@apollo/client/react/components';
+import { Trash2 as IconTrash } from '@styled-icons/feather/Trash2';
+import { FormattedMessage, useIntl } from 'react-intl';
+
+import { i18nGraphqlException } from '../../lib/errors';
+import { API_V2_CONTEXT, gqlV2 } from '../../lib/graphql/helpers';
+
+import ConfirmationModal from '../ConfirmationModal';
+import StyledButton from '../StyledButton';
+import { Span } from '../Text';
+import { TOAST_TYPE, useToasts } from '../ToastProvider';
+
+const deleteExpenseMutation = gqlV2/* GraphQL */ `
+  mutation DeleteExpense($id: String!) {
+    deleteExpense(expense: { id: $id }) {
+      id
+    }
+  }
+`;
+
+const removeExpenseFromCache = (cache, { data: { deleteExpense } }) => {
+  cache.modify({
+    fields: {
+      expenses(existingExpenses, { readField }) {
+        if (!existingExpenses?.nodes) {
+          return existingExpenses;
+        } else {
+          return {
+            ...existingExpenses,
+            totalCount: existingExpenses.totalCount - 1,
+            nodes: existingExpenses.nodes.filter(expense => deleteExpense.id !== readField('id', expense)),
+          };
+        }
+      },
+    },
+  });
+};
+
+const DeleteExpenseButton = ({ expense, onDelete, buttonProps, isDisabled, onModalToggle }) => {
+  const [hasDeleteConfirm, setDeleteConfirm] = React.useState(false);
+  const { addToast } = useToasts();
+  const intl = useIntl();
+  const showDeleteConfirm = isOpen => {
+    setDeleteConfirm(isOpen);
+    onModalToggle?.(isOpen);
+  };
+
+  return (
+    <React.Fragment>
+      <StyledButton
+        buttonStyle="dangerSecondary"
+        data-cy="delete-expense-button"
+        disabled={isDisabled}
+        {...buttonProps}
+        onClick={() => showDeleteConfirm(true)}
+      >
+        <IconTrash size="1em" />
+        <Span ml="6px">
+          <FormattedMessage id="actions.delete" defaultMessage="Delete" />
+        </Span>
+      </StyledButton>
+      {hasDeleteConfirm && (
+        <Mutation mutation={deleteExpenseMutation} context={API_V2_CONTEXT} update={removeExpenseFromCache}>
```

Comment:
The [component API of apollo is deprecated](https://www.apollographql.com/docs/react/api/react/components/) and maybe we should use hooks (here `useMutation`) for new development? 🤔

Reply:
We should indeed. But here it's not a new development, I just moved existing code around.